### ISSUE TITLE
[HW3-kcov] Clarification about 'how to design enum-safe probes'

### ISSUE THREAD

**Issue by Allgot** [swtv-kaist/cs458-spring23#21]
Hi! I have a question about designing probes, which are resistant to enums in C/C++.

As the professor said during the lecture, in C/C++, every values should be calculated in complie-time, but not runtime.
Therefore, we have to carefully insert probes not to violate this requirements.

However, I ran out of my own ideas and I feel frustrated... DX

At first, I tried to **count** the case which the conditional operatiors are nested in enum definitions. But it faield(my implementation failed to count such cases exhaustively), so I took another approach: if I met enum definitions, just 'ignore' nested enum definitions. Again, if failed.. I couldn't think up when to restore...(i.e. when to stop ignoring)

Could anyone give me some hints? Any tiny one is okay...!

**Comment by YoungseokCh** [comment]
> At first, I tried to count the case which the conditional operatiors are nested in enum definitions. But it faield(my implementation failed to count such cases exhaustively), so I took another approach: if I met enum definitions, just 'ignore' nested enum definitions. Again, if failed.. I couldn't think up when to restore...(i.e. when to stop ignoring)

Could you elaborate on this part with a detailed example? It's not very clear to me what kind of problems you're having.

**Comment by YoungseokCh** [comment]
Note that you can ignore the conditional operators in enumerator definitions because every value inside an enum is constant and determined in compile-time.

**Comment by jaehyun1ee** [comment]
Maybe consider only the statements inside a function declaration (definition) 🙂

**Comment by Allgot** [comment]
First of all, I'm so sorry for the pool description. And I've resolved it through a few of assumptions, thanks to you two!

And let me give some details about my trial and error for others, if exist (more about lessons what I've learned.. ):
counting the number of such cases can be too complicated but even less accurate;
to ignore such cases, of course I can make a new function... but Clang provides bunch of pre-defined operations, so it would be great to search first.
